app/models/ownership.rb

```diff
@@ -12,7 +21,54 @@ def self.by_indexed_gem_name
       .order("rubygems.name ASC")
   end
 
+  def valid_confirmation_token?
+    token_expires_at > Time.zone.now
+  end
+
+  def generate_confirmation_token
+    self.token = SecureRandom.hex(20).encode("UTF-8")
+    self.token_expires_at = Time.zone.now + Gemcutter::OWNERSHIP_TOKEN_EXPIRES_AFTER
+  end
+
+  def confirm_ownership!
+    update(confirmed_at: Time.current)
+  end
+
+  def confirmed?
+    confirmed_at.present?
+  end
+
+  def unconfirmed?
+    !confirmed?
+  end
+
+  def notify_owner_removed
+    rubygem.notifiable_owners.each do |notified_user|
+      Mailer.delay.owner_removed(user_id, notified_user.id, rubygem_id)
+    end
+  end
+
+  def notify_owner_added
+    rubygem.notifiable_owners.each do |notified_user|
+      Mailer.delay.owner_added(user_id, notified_user.id, rubygem_id)
+    end
+  end
```

Comment:
Looks like `notify_owner_added` and `notify_owner_removed` don't have to be part of public interface.

`safe_destroy` could be moved to private as well. If there are existing tests, use `destroy_and_notify` to test the same behaviour.